src/simulators/statevector/qubitvector_thrust.hpp

```diff
@@ -79,6 +89,722 @@ using indexes_t = std::unique_ptr<uint_t[]>;
 template <size_t N> using areg_t = std::array<uint_t, N>;
 template <typename T> using cvector_t = std::vector<std::complex<T>>;
 
+//==================================
+// parameters for gate kernels
+//==================================
+template <typename data_t>
+struct GateParams {
+  thrust::complex<data_t>* buf_;
+  uint_t* offsets_;
+  thrust::complex<double>* matrix_;
+  uint_t* params_;
+  uint_t gid_;
+  uint_t lmask_;
+};
+
+//========================================
+//  base class of gate functions
+//========================================
+class GateFuncBase
+{
+public:
+  GateFuncBase(void)
+  {
+  }
+
+  virtual bool IsDiagonal(void)
+  {
+    return false;
+  }
+  virtual int NumControlBits(void)
+  {
+    return 0;
+  }
+  virtual int ControlMask(void)
+  {
+    return 1;
+  }
+  virtual bool Reduction(void)
+  {
+    return false;
+  }
+
+  virtual const char* Name(void)
+  {
+    return "base function";
+  }
+};
+
+
+//=============================================================
+//    virtual buffer class
```

Comment:
I think these thrust helper classes should be moved to a separate header file and included. It would make the code easier to read (ie `QubitVectorBuffer`, `QubitVctorDeviceBuffer`, `QubitVectorHostBuffer`, `QubitVectorChunkContainer).

Some simple usage comments on the methods of the classes would be useful to describe what they do, and what arguments are)